docs/content/preview/architecture/transactions/read-restart-error.md

```diff
@@ -15,26 +15,68 @@ rightNav:
+
+  ![Read Restart Error](/images/architecture/txn/read_restart_error.svg)
+
+  In the above setup, there are three tablet servers, namely, `TSERVER 1`, `TSERVER 2`, and `TSERVER 3`. The data for the tokens table is hosted on servers TSERVER 2, and TSERVER 3. The query layer on TSERVER 1 is serving the SQL requests.
+
+  Moreover, TSERVER 3's clock is running 5 units of time ahead of TSERVER 2's clock because of clock skew.
+
+1. First, an INSERT of token 29 is issued to YSQL on TSERVER 1.
+2. This INSERT is routed to the tablet hosted on TSERVER 3. The operation picks a commit time of `T2=103` even though the "true time" is `98` (step 2).
+3. TSERVER 1 acknowledges the INSERT.
+4. Now, that the INSERT command is complete, the SELECT command is issued.
+5. TSERVER 1 starts a distributed read. TSERVER 1 reads data from multiple shards on different physical YB-TServers in the cluster, namely TSERVER 2 and TSERVER 3. The read point that defines the snapshot of the database at which the data will be read, is picked on TSERVER 2 based on the safe time of that YB-TServer, namely `T1=101`.
+6. TSERVER 1 gets back token 17 in the result set.
+7. TSERVER 1 now issues a read operation to TSERVER 3 to retrieve records from the second shard. However, since the read time `T1=101` is less than the commit time of the prior insert `T2=103`, the record is not part of the read snapshot.
+8. Thus, token 29 is omitted from the result set returned.
+
+  More generally, this anomaly occurs whenever the commit time of a prior write operation `T2` is higher than the read time `T1` of a later read operation, thus violating (distributed) consistency.
+
+How does YugabyteDB mitigate this clock skew anomaly?
+
+* First, note that the clock skew between all nodes in the cluster is always in a [max_clock_skew_usec](../../../reference/configuration/yb-tserver/#max-clock-skew-usec) bound due to clock synchronization algorithms.
+* For records with a commit timestamp later than `T1` + `max_clock_skew`, the database can be sure that these records were written after the read was issued. But for records with a commit timestamp between `T1` and `T1` + `max_clock_skew`, the tablet servers can find themselves in an ambiguous situation, such as the following:
+
+  * A read operation must observe all the commits that were acknowledged before the read was issued, even when the read time is picked on a node whose clock runs behind the node that committed the prior write. NOTE: the commit happened before the read as perceived by the user and not necessarily according to the machine clocks.
+
+  * A read operation must return a consistent snapshot of data, i.e. the read cannot simply advance its read time on observing a record with higher timestamp since the read already returned records from an older snapshot thus far. Therefore, the read must be restarted from the beginning with the advanced timestamp. Thus, the name read restart error.
```

Comment:
```suggestion
  * A read operation must return a consistent snapshot of data. That is, the read cannot simply advance its read time on observing a record with higher timestamp since the read already returned records from an older snapshot thus far. Therefore, the read must be restarted from the beginning with the advanced timestamp. Thus, the name read restart error.
```